src/hosting/api.ts

```diff
@@ -85,6 +85,85 @@ export interface Channel {
   labels: { [key: string]: string };
 }
 
+enum VersionStatus {
+  // The default status; should not be intentionally used.
+  VERSION_STATUS_UNSPECIFIED = "VERSION_STATUS_UNSPECIFIED",
+  // The version has been created, and content is currently being added to the
+  // version.
+  CREATED = "CREATED",
+  // All content has been added to the version, and the version can no longer be
+  // changed.
+  FINALIZED = "FINALIZED",
+  // The version has been deleted.
+  DELETED = "DELETED",
+  // The version was not updated to `FINALIZED` within 12&nbsp;hours and was
+  // automatically deleted.
+  ABANDONED = "ABANDONED",
+  // The version is outside the site-configured limit for the number of
+  // retained versions, so the version's content is scheduled for deletion.
+  EXPIRED = "EXPIRED",
+  // The version is being cloned from another version. All content is still
+  // being copied over.
+  CLONING = "CLONING",
+}
+
+// TODO: define ServingConfig.
+enum ServingConfig {}
+
+export interface Version {
+  // The unique identifier for a version, in the format:
+  // `sites/<site-name>/versions/<versionID>`
+  name: string;
+
+  // The deploy status of a version.
+  status: VersionStatus;
+
+  // The configuration for the behavior of the site.
+  config: ServingConfig;
+
+  // The labels used for extra metadata and/or filtering.
+  labels: Map<string, string>;
+
+  // The time at which the version was created.
+  readonly createTime: string;
+
+  // Identifies the user who created the version.
+  readonly createUser: ActingUser;
+
+  // The time at which the version was `FINALIZED`.
+  readonly finalizeTime: string;
+
+  // Identifies the user who `FINALIZED` the version.
+  readonly finalizeUser: ActingUser;
+
+  // The time at which the version was `DELETED`.
+  readonly deleteTime: string;
+
+  // Identifies the user who `DELETED` the version.
+  readonly deleteUser: ActingUser;
+
+  // The total number of files associated with the version.
+  readonly fileCount: number;
+
+  // The total stored bytesize of the version.
+  readonly versionBytes: number;
+}
+
+/**
+ * normalizeName normalizes a name given to it. Most useful for normalizing
+ * user provided names. This removes any `/` characters and replaces them with
+ * dashes (`-`).
+ * @param s the name to normalize.
+ * @return the normalized name.
+ */
+export function normalizeName(s: string): string {
```

Comment:
I think we should also normalize `:`, and maybe `_` as well (since underscores aren't technically valid subdomain names).

Reply:
done.